### PR TITLE
Allow hlint to function as a CC engine

### PR DESCRIPTION
**Pull request by pbrisbin**
- Add `--json-cc` flag to output in in accordance with the CC SPEC[1].
- Add `cc/` directory with the machinery required to run as a codeclimate
  engine. The only bit that couldn't be isolated in this directory is
   the .dockerignore which must live in the directory you're building
   from. 
- Add a .codeclimate.yml so we can test the engine on ourselves.

[1]: https://github.com/codeclimate/spec/blob/master/SPEC.md

This should replace pbrisbin/codeclimate-hlint and address the requested features around configurability. This version of the engine runs the hlint binary normally, so passing additional flags, accessing project-local configuration files, etc, should all Just Work.

Try it out:

    cc/build
    codeclimate analyze

See https://github.com/codeclimate/codeclimate#installation

This is meant as a jumping-off point: please let me know if there's a better home for `CC.hs`, a better name for the `--json-cc` flag, or a better place to put the `cc/` Docker machinery.

One other question: I notice `stack.yml` is git-ignored. That file will have to exist for the docker image to build, and code climate maintainers will be doing the image build out of a fresh checkout, so I'm not sure if we should commit one, or put one in `cc/` that we can use as part of the `cc/build` process?

### CODE REVIEW

**Comment by ndmitchell** [comment]
Thanks for the code! I'll review properly in the next few days.

For `stack.yaml` I already test that `stack init` generates a working `stack.yaml`, which is why I don't have one in the directory. Could the code climate scripts just call `stack init`? I'm generally not a fan of putting dates in a repo that need incrementing on some schedule.

**Comment by pbrisbin** [comment]
> Could the code climate scripts just call stack init?

That's fine with me, except it doesn't seem to work:

```
Resolver 'lts-9.2' does not have all the packages to match your requirements.
    haskell-src-exts-util not found
        - hlint requires >=0.2.1.2
```

Everything is fine if I make a `stack.yaml` that specifies the `extra-deps`.

AFAICT, this affects master hlint overall, not just my attempts at building.

**Comment by pbrisbin** [comment]
Oh by the way, if you'd prefer, it'd be reasonable to reduce this PR to only the `--json-cc` addition. The machinery under `cc/` could rename in my `codeclimate-hlint` project. Please let me know if that'd be a better first step.

**Comment by ndmitchell** [comment]
Given currently a stack is required, we should add it - at least for now. I will aim to remove it in future though. I'm happy to review all at once.

File: cc/build [hunk 1]
```diff
@@ -0,0 +1,6 @@
+#!/bin/sh
```
Comment (ndmitchell) [inline cc/build]: Is build the convention? I tend to prefer build.sh so people know it's a shell script. However, I'm not super fussed.

File: cc/engine.hs [hunk 2]
```diff
@@ -0,0 +1,83 @@
+{-# LANGUAGE OverloadedStrings #-}
+module Main where
```
Comment (ndmitchell) [inline cc/engine.hs]: Please add an explicit export list of 'main'

File: cc/engine.hs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+import Data.Maybe (fromMaybe)
+import System.Process (callProcess)
+import System.IO (hPutStrLn, stderr)
+
+import qualified Data.ByteString.Lazy.Char8 as C8
+
+data Config = Config
+    { cIncludePaths :: [FilePath]
+    , cExtraFlags :: [String]
+    }
```
Comment (ndmitchell) [inline cc/engine.hs]: Is this a config format we are defining? Any reason not to reuse something that already exists? Or is this a code climate format?

File: cc/engine.hs [hunk 4]
```diff
@@ -0,0 +1,83 @@
+{-# LANGUAGE OverloadedStrings #-}
```
Comment (ndmitchell) [inline cc/engine.hs]: I'd prefer Engine with a leading capital

File: hlint.cabal [hunk 5]
```diff
@@ -126,3 +133,19 @@ executable hlint
     ghc-options:        -rtsopts
     if flag(threaded)
         ghc-options:    -threaded
+
+executable cc-engine
```
Comment (ndmitchell) [inline hlint.cabal]: Could we roll this executable into Hlint - eg have "Hlint cc" do the necessary magic?

Comment (ndmitchell) [reply hlint.cabal]: Assuming we do need a separate executable, would it make more sense to have cc/hlint-cc.cabal? I don't see that many benefits to keeping it in the main one.

File: src/CC.hs [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        , T.pack to
+        , "```"
+        ]
+
+    categories _ = ["Style"]
+
+    points Ignore = 0
+    points Suggestion = basePoints
+    points Warning = 5 * basePoints
+    points Error = 10 * basePoints
```
Comment (ndmitchell) [inline src/CC.hs]: It seems like these points are markers of difficulty rather than importance - so are these numbers sensible?

File: src/CC.hs [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+printIssue :: Issue -> IO ()
+printIssue = C8.putStrLn . (<> "\0") . encode
+
+-- | Convert an hlint @'Idea'@ to a datatype more easily serialized for CC
+fromIdea :: Idea -> Issue
+fromIdea Idea{..} = Issue
+    { issueType = "issue"
+    , issueCheckName = "HLint/" <> T.pack (camelize ideaHint)
+    , issueDescription = T.pack ideaHint
+    , issueContent = content ideaFrom ideaTo
```
Comment (ndmitchell) [inline src/CC.hs]: You are dropping ideaNote which should definitely end up in the content - those are warnings about when refactorings have subtle side conditions.

File: src/CC.hs [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+import qualified Data.Text as T
+import qualified Data.ByteString.Lazy.Char8 as C8
+
+import Idea (Idea(..), Severity(..))
+
+data Issue = Issue
+    { issueType :: Text
+    , issueCheckName :: Text
+    , issueDescription :: Text
+    , issueContent :: Text -- TODO: Markdown type?
```
Comment (ndmitchell) [inline src/CC.hs]: I don't think a markdown type gives enough benefits here.

File: src/CC.hs [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+    , issueCategories :: [Text]
+    , issueLocation :: Location
+    , issueRemediationPoints :: Int
+    }
+
+data Location = Location FilePath Position Position
```
Comment (ndmitchell) [inline src/CC.hs]: I wonder if defining srcSpanJSON makes sense, so you avoid the orphan instance and avoid a separate location type. Would avoid multiple conversions and let you put SrcSpan in Issue?

File: src/CmdLine.hs [hunk 10]
```diff
@@ -174,6 +175,7 @@ mode = cmdArgsMode $ modes
         ,cmdCppSimple = nam_ "cpp-simple" &= help "Use a simple CPP (strip # lines)"
         ,cmdCppAnsi = nam_ "cpp-ansi" &= help "Use CPP in ANSI compatibility mode"
         ,cmdJson = nam_ "json" &= help "Display hint data as JSON"
+        ,cmdJsonCC = nam_ "json-cc" &= help "Display hint data as Code Climate Issues JSON"
```
Comment (ndmitchell) [inline src/CmdLine.hs]: Should the flag just be --code-climate? The important thing is it's in code climate format. The fact code climate format is JSON is just a detail.

File: cc/build (hunk 1 above)
Comment (pbrisbin) [reply cc/build]: I'm happy to name it whatever you'd like. My personally style is to never use extensions with executables. Callers shouldn't know or care, and we should be free to change, the implementation language without changing the name. Would you like me to rename it `build.sh`?

File: cc/engine.hs (hunk 3 above)
Comment (pbrisbin) [reply cc/engine.hs]: This is (our materialization of) the defined `/config.json` format we operate with as a CC Engine. I tried to explain how it works a little bit in the comment down with the `ToJSON` instance.

File: hlint.cabal (hunk 5 above)
Comment (pbrisbin) [reply hlint.cabal]: I thought it was cleaner to keep a separate executable, and only build it on a flag. IMO, users (and developers) of `hlint` directly shouldn't really be exposed to or have to think about the CC cruft daty-to-day. Had I thought of it, I would've done `cc/engine.cabal`, so I'm in favor of that.

That said, if you prefer `hlint cc` just let me know, I see now reason that wouldn't work.

File: src/CC.hs (hunk 6 above)
Comment (pbrisbin) [reply src/CC.hs]: I would love you input on that. The comment down with `basePoints` describes how it's meant to work, using that could you recommend rough numbers for the various severities?

File: src/CC.hs (hunk 7 above)
Comment (pbrisbin) [reply src/CC.hs]: I didn't realize that; could you offer a rough template of what content should look like, taking that into account?

File: src/CC.hs (hunk 9 above)
Comment (pbrisbin) [reply src/CC.hs]: You're suggesting making a `ToJSON SrcSpan` and using it directly here? To me it feels wrong to make *the* JSON instance for anything conform CC's definition, so I'd lean towards a separate type anyway (even if just a `newtype` to allow a specific instance). Additionally, `SrcSpan -> Location` also handles relativizing the file-path (albeit naively at the moment). Taken together, these make the `Location` type wortwhile IMO.

Those are my reasons, but again it's totally your call if you'd like to do something else.

File: src/CmdLine.hs (hunk 10 above)
Comment (pbrisbin) [reply src/CmdLine.hs]: `--code-climate` works for me. `--cc` would also be OK with me, I'm not sure if there's risk in the brevity, WDYT?

**Comment by pbrisbin** [comment]
@ndmitchell just a friendly bump on this. Any other changes you'd like to see?

**Comment by ndmitchell** [comment]
Merged! Sorry it took so long - life went crazy :)

**Comment by pbrisbin** [comment]
No worries, thank you! I'll touch base with the CC folks next week and get this new version into their system.

**Comment by ndmitchell** [comment]
I just released 2.0.10, so you have a released version to build upon.

**Pull request merged**